book/custom_commands.md

```diff
@@ -252,187 +444,350 @@ def greet [
+
+```text
+Greet guests along with a VIP
+
+Use for birthdays, graduation parties,
+retirements, and any other event which
+celebrates an event # for a particular
+person.
+
+Category: default
+
+This command:
+- does not create a scope.
+- is not a built-in command.
+- is not a subcommand.
+- is not part of a plugin.
+- is a custom command.
+- is not a keyword.
+
+Usage:
+  > vip-greet <vip>
+
+
+Flags:
+
+
+  -h, --help - Display the help message for this command
+
+Signatures:
+
+  <any> | vip-greet[ <string>] -> <any>
+
+Parameters:
+
+  vip: <string> The special guest
+  ...rest: <string> The other guests
+```
+
+Notice that the comments on the lines immediately before the `def` statement become a description of the command in the help system. Multiple lines of comments can be used. The first line (before the blank-comment line) becomes the Help `description`. This information is also shown when tab-completing commands.
+
+The remaining comment lines become its `extra_description` in the help data.
+
+::: tip
+Run:
+
+```nu
+scope commands
+| where name == 'vip-greet'
+| wrap help
+```
 
-::: warning Note
+This will show the Help _record_ that Nushell creates.
+:::
+
+The comments following the parameters become their description. Only a single-line comment is valid for parameters.
+
+::: tip Note
 A Nushell comment that continues on the same line for argument documentation purposes requires a space before the ` #` pound sign.
 :::
 
-Now, if we run `help greet`, we're given a more helpful help text:
+## Changing the Environment in a Custom Command
+
+Normally, environment variable definitions and changes are [_scoped_ within a block](./environment.html#scoping). This means that changes to those variables are lost when they go out of scope at the end of the block, including the block of a custom command.
 
+```nu
+def foo [] {
+    $env.FOO = 'After'
+}
+
+$env.FOO = "Before"
+foo
+$env.FOO
+# => Before
 ```
-A greeting command that can greet the caller
 
-Usage:
-  > greet <name> {flags}
+However, a command defined using [`def --env`](/commands/docs/def.md) or [`export def --env`](/commands/docs/export_def.md) (for a [Module](modules.md)) will preserve the environment on the caller's side:
 
-Parameters:
-  <name> The name of the person to greet
+```nu
+def --env foo [] {
+    $env.FOO = 'After'
+}
 
-Flags:
-  -h, --help: Display this help message
-  -a, --age <integer>: The age of the person
+$env.FOO = "Before"
+foo
+$env.FOO
+# => After
 ```
 
-## Pipeline Output
+### Changing Directories in a Custom Command
+
+Likewise, changing the directory using the `cd` command results in a change of the `$env.PWD` environment variable. This means that directory changes (the `$env.PWD` variable) will also be reset when a custom command ends. The solution, as above, is to use `def --env` or `export def --env`.
+
+```nu
+def --env go-home [] {
+  cd ~
+}
+
+cd /
+go-home
+pwd
+# => Your home directory
+```
+
+## Custom Commands and Pipelines
```

Comment:
I think this section should be moved above the "Input-Output Signature" section, since it's not clear what that section is referring to without the context of this section